arviz/sel_utils.py

```diff
@@ -0,0 +1,224 @@
+
+    Parameters
+    ----------
+    data : xarray.Dataset
+        Posterior data in an xarray
+
+    var_names : iterator of strings (optional)
+        Should be a subset of data.data_vars. Defaults to all of them.
+
+    combined : bool
+        Whether to combine chains or leave them separate
+
+    skip_dims : set
+        dimensions to not iterate over
+
+    reverse_selections : bool
+        Whether to reverse selections before iterating.
+
+    Returns
+    -------
+    Iterator of (var_name: str, selection: dict(str, any))
+        The string is the variable name, the dictionary are coordinate names to values,.
+        To get the values of the variable at these coordinates, do
+        ``data[var_name].sel(**selection)``.
+    """
+    if skip_dims is None:
+        skip_dims = set()
+
+    if combined:
+        skip_dims = skip_dims.union({"chain", "draw"})
+    else:
+        skip_dims.add("draw")
+
+    if var_names is None:
+        if isinstance(data, xr.Dataset):
+            var_names = list(data.data_vars)
+        elif isinstance(data, xr.DataArray):
+            var_names = [data.name]
+            data = {data.name: data}
+
+    for var_name in var_names:
+        if var_name in data:
+            new_dims = _dims(data, var_name, skip_dims)
+            vals = [purge_duplicates(data[var_name][dim].values) for dim in new_dims]
+            dims = _zip_dims(new_dims, vals)
+            if reverse_selections:
+                dims = reversed(dims)
+
+            for selection in dims:
+                yield var_name, selection
+
+
+def xarray_var_iter(data, var_names=None, combined=False, skip_dims=None, reverse_selections=False):
+    """Convert xarray data to an iterator over vectors.
+
+    Iterates over each var_name and all of its coordinates, returning the 1d
+    data.
+
+    Parameters
+    ----------
+    data : xarray.Dataset
+        Posterior data in an xarray
+
+    var_names : iterator of strings (optional)
+        Should be a subset of data.data_vars. Defaults to all of them.
+
+    combined : bool
+        Whether to combine chains or leave them separate
+
+    skip_dims : set
+        dimensions to not iterate over
+
+    reverse_selections : bool
+        Whether to reverse selections before iterating.
+
+    Returns
+    -------
+    Iterator of (str, dict(str, any), np.array)
+        The string is the variable name, the dictionary are coordinate names to values,
+        and the array are the values of the variable at those coordinates.
+    """
+    data_to_sel = data
+    if var_names is None and isinstance(data, xr.DataArray):
+        data_to_sel = {data.name: data}
+
+    for var_name, selection in xarray_sel_iter(
+        data,
+        var_names=var_names,
+        combined=combined,
+        skip_dims=skip_dims,
+        reverse_selections=reverse_selections,
+    ):
+        yield var_name, selection, data_to_sel[var_name].sel(**selection).values
```

Comment:
selection should be tuple with `iloc` and `loc` selection methods, right?

Reply:
that sounds good, how about `yield var_name, selection, iselection, data_to_sel...`?

Reply:
that option sounds good

Reply:
I was looking into this and I realized that having indexes is currently close to impossible, I think it would be better to move this to an issue and also discuss what we want to be possible in a lab meeting.

One of the first operations done is `dataset.sel()`, so the correct index information is lost before calling `xarray_sel_iter`.